File: akka-docs/src/test/java/jdocs/stream/javadsl/cookbook/RecipeSplitter.java
/*
 * Copyright (C) 2016-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.stream.javadsl.cookbook;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.testkit.javadsl.TestKit;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutionException;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

public class RecipeSplitter {
  private static ActorSystem system;

  @Test
  public void simpleSplit() throws ExecutionException, InterruptedException {
    // #Simple-Split
    // Sample Source
    Source<String, NotUsed> source = Source.from(Arrays.asList("1-2-3", "2-3", "3-4"));

    CompletionStage<List<Integer>> ret =
        source
            .map(s -> Arrays.asList(s.split("-")))
            .mapConcat(f -> f)
            // Sub-streams logic
            .map(s -> Integer.valueOf(s))
            .runWith(Sink.seq(), system);

    // Verify results
    List<Integer> list = ret.toCompletableFuture().get();
    assert list.equals(Arrays.asList(1, 2, 3, 2, 3, 3, 4));
    // #Simple-Split
  }

  @Test
  public void splitAggregate() throws ExecutionException, InterruptedException {
    // #Aggregate-Split
    // Sample Source
    Source<String, NotUsed> source = Source.from(Arrays.asList("1-2-3", "2-3", "3-4"));

    CompletionStage<List<Integer>> ret =
        source
            .map(s -> Arrays.asList(s.split("-")))
            // split all messages into sub-streams
            .splitWhen(a -> true)
            // now split each collection
            .mapConcat(f -> f)
            // Sub-streams logic
            .map(s -> Integer.valueOf(s))
            // aggregate each sub-stream
            .reduce((a, b) -> a + b)
            // and merge back the result into the original stream
            .mergeSubstreams()
            .runWith(Sink.seq(), system);

    // Verify results
    List<Integer> list = ret.toCompletableFuture().get();
    assert list.equals(Arrays.asList(6, 5, 7));
    // #Aggregate-Split
  }

  @BeforeClass
  public static void setup() throws Exception {
    system = ActorSystem.create();
  }

  @AfterClass
  public static void teardown() throws Exception {
    TestKit.shutdownActorSystem(system);
  }
}

File: akka-docs/src/test/scala/docs/stream/cookbook/RecipeSplitter.scala
/*
 * Copyright (C) 2016-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.stream.cookbook

import akka.NotUsed
import akka.actor.ActorSystem
import akka.stream.scaladsl.{ Sink, Source }
import org.scalatest.BeforeAndAfterAll
import org.scalatest.concurrent.ScalaFutures
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

class RecipeSplitter extends AnyWordSpec with BeforeAndAfterAll with Matchers with ScalaFutures {

  implicit val system: ActorSystem = ActorSystem("Test")

  "Splitter" should {
    " simple split " in {

      // #Simple-Split
      // Sample Source
      val source: Source[String, NotUsed] = Source(List("1-2-3", "2-3", "3-4"))

      val ret = source
        .map(s => s.split("-").toList)
        .mapConcat(identity)
        // Sub-streams logic
        .map(s => s.toInt)
        .runWith(Sink.seq)

      // Verify results

      ret.futureValue should be(Vector(1, 2, 3, 2, 3, 3, 4))
      // #Simple-Split

    }

    " aggregate split" in {

      // #Aggregate-Split
      // Sample Source
      val source: Source[String, NotUsed] = Source(List("1-2-3", "2-3", "3-4"))

      val result = source
        .map(s => s.split("-").toList)
        // split all messages into sub-streams
        .splitWhen(a => true)
        // now split each collection
        .mapConcat(identity)
        // Sub-streams logic
        .map(s => s.toInt)
        // aggregate each sub-stream
        .reduce((a, b) => a + b)
        // and merge back the result into the original stream
        .mergeSubstreams
        .runWith(Sink.seq);

      // Verify results
      result.futureValue should be(Vector(6, 5, 7))
      // #Aggregate-Split

    }

  }

  override protected def afterAll(): Unit = system.terminate()

}
